File: src/main/scala/org/joisen/chapter11/TopNExample.scala
package org.joisen.chapter11

import org.apache.flink.streaming.api.scala._
import org.apache.flink.table.api.Table
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment

/**
 * @Author Joisen
 * @Date 2022/11/28 10:55
 * @Version 1.0
 */
object TopNExample {
  def main(args: Array[String]): Unit = {
    // 1 创建表环境
    // 1.1 直接基于流执行环境创建
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    val tableEnv: StreamTableEnvironment = StreamTableEnvironment.create(env)


    // 1 在创建表的DDL中指定时间属性
    tableEnv.executeSql("create table eventTable(" +
      "uid string," +
      "url string," +
      "ts bigint," +
      " et as TO_TIMESTAMP( FROM_UNIXTIME(ts/1000))," +
      " WATERMARK FOR et AS et - INTERVAL '2' SECOND" +
      ") with(" +
      " 'connector' = 'filesystem', " +
      " 'path' = 'input/clicks.txt', " +
      " 'format' = 'csv' " +
      ")")

    // TOP N  选取活跃度最大的前2个用户
    // 1. 进行分组聚合统计，计算每个用户目前的访问量
    val urlCountTable: Table = tableEnv.sqlQuery("select uid, count(url) as cnt from eventTable group by uid")
    tableEnv.createTemporaryView("urlCountTable", urlCountTable)
    // 提取count值最大的前2个用户
    val top2ResTbl: Table = tableEnv.sqlQuery(
      """
        |select uid, cnt, row_num
        |from(
        | select *, row_number() over(
        |   order by cnt desc
        | ) as row_num
        | from urlCountTable
        |)
        |where row_num <= 2
        |""".stripMargin)
    tableEnv.toChangelogStream(top2ResTbl).print()
    env.execute()


  }
}

File: src/main/java/org/joisen/java/chapter11/TopNExample.java
package org.joisen.java.chapter11;

import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;

/**
 * @Author Joisen
 * @Date 2022/12/14 15:31
 * @Version 1.0
 */
public class TopNExample {
    public static void main(String[] args) throws Exception{
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);
        // 1 在创建表的DDL中直接定义时间属性
        String createDDL = "create table clickTable (" +
                " usr string," +
                " url string,"+
                " ts bigint, " +
                " et as TO_TIMESTAMP(FROM_UNIXTIME(ts/1000)), " +
                " WATERMARK for et as et - INTERVAL '1' second" +
                " ) with( " +
                " 'connector' = 'filesystem', " +
                " 'path' = 'input/clicks.txt', " +
                " 'format' = 'csv' )";
        tableEnv.executeSql(createDDL);

        // 普通TOPN， 选取当前所有用户中浏览量最大的2个
        Table topNResTbl = tableEnv.sqlQuery("select usr, cnt, row_num " +
                "from(" +
                " select *, row_number() over(" +
                " order by cnt desc" +
                " ) as row_num " +
                " from (select usr, count(url) as cnt from clickTable group by usr )" +
                ") where row_num <= 2");

//        tableEnv.toChangelogStream(topNResTbl).print("top 2:");

        // 窗口TOP N 统计一段时间内的（前2）活跃用户
        String subQuery = "select usr, count(url) as cnt, window_start, window_end" +
                " from table(" +
                "   tumble(table clickTable, descriptor(et), interval '10' second)" +
                ")" +
                " group by usr, window_start, window_end ";

        Table windowTopNResTbl = tableEnv.sqlQuery("select usr, cnt, row_num, window_end " +
                "from (" +
                "   select *, row_number() over(" +
                "       partition by window_start, window_end " +
                "       order by cnt desc" +
                "   ) as row_num " +
                "   from (" + subQuery + " )" +
                ") where row_num <= 2");
        tableEnv.toDataStream(windowTopNResTbl).print("window_top 2:");


        env.execute();
    }
}
